### PR TITLE
Add metrics for retries by S3RetryingInputStream

### PR DESCRIPTION
**Pull request by ywangd**
This PR exposes retries in S3RetryingInputStream as metrics for easier observability. At the class API level, retries can happen when either opening an input stream and reading from an input stream. Retry reading from an input stream internally can also retry re-opening the input stream. All these retries are counted under the retries for reading since the higher API usage is a read instead of open.

The list of new metrics are:
* `es.repositories.s3.input_stream.retry.event.total` - Number of times a retry cycle has been triggered.
* `es.repositories.s3.input_stream.retry.success.total` - Number of a times a retry cycle has been successfully completed. This should match the above metric in numbers. Otherwise it indicates there are threads stuck in infinite retries.
* `es.repositories.s3.input_stream.retry.attempts.histogram` - Number of attempts to complete a retry cycle successfully.

Relates: https://github.com/elastic/elasticsearch/pull/103300#discussion_r1444125047
Relates: ES-7666

### CODE REVIEW

**Comment by ywangd** [comment]
Note to reviewer: There are quite a few cascading changes due to the refactor of `RepositoriesMetrics`. The essence of the changes is rather small and can be found in [S3RetryingInputStream](https://github.com/elastic/elasticsearch/pull/105600/files#diff-dfd144850a491a4aa42eb5c1e0de797174a21d3f56633f85e45af81153bf408b)

**Comment by ywangd** [comment]
Failure is https://github.com/elastic/elasticsearch/issues/103588

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RepositoriesMetrics.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    LongCounter retryStartedCounter,
+    LongCounter retryCompletedCounter,
+    LongHistogram retryHistogram
+) {
+
+    public static S3RepositoriesMetrics NOOP = new S3RepositoriesMetrics(RepositoriesMetrics.NOOP);
+
+    public static final String METRIC_RETRY_EVENT_TOTAL = "es.repositories.s3.input_stream.retry.event.total";
+    public static final String METRIC_RETRY_SUCCESS_TOTAL = "es.repositories.s3.input_stream.retry.success.total";
+    public static final String METRIC_RETRY_ATTEMPTS_HISTOGRAM = "es.repositories.s3.input_stream.retry.attempts.histogram";
```
Comment (ywangd) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RepositoriesMetrics.java:26]: @idegtiarenko I renamed the first two metrics. Please let me know if they work better for you. Thanks!

**Review by idegtiarenko — APPROVED** [review]

**Comment by ywangd** [comment]
@elasticmachine update branch

**Pull request merged**